Mathlib/Order/Heyting/Basic.lean

```diff
@@ -0,0 +1,1363 @@
+  sorry
+  -- pi_instance
+  -- exact fun a b c => forall_congr' fun i => sdiff_le_iff
+#align pi.generalized_coheyting_algebra Pi.generalizedCoheytingAlgebra
+
+end GeneralizedCoheytingAlgebra
+
+section HeytingAlgebra
+
+variable [HeytingAlgebra α] {a b c : α}
+
+@[simp]
+theorem himp_bot (a : α) : a ⇨ ⊥ = aᶜ :=
+  HeytingAlgebra.himp_bot _
+#align himp_bot himp_bot
+
+@[simp]
+theorem bot_himp (a : α) : ⊥ ⇨ a = ⊤ :=
+  himp_eq_top_iff.2 bot_le
+#align bot_himp bot_himp
+
+theorem compl_sup_distrib (a b : α) : (a ⊔ b)ᶜ = aᶜ ⊓ bᶜ := by
+  simp_rw [← himp_bot, sup_himp_distrib]
+#align compl_sup_distrib compl_sup_distrib
+
+@[simp]
+theorem compl_sup : (a ⊔ b)ᶜ = aᶜ ⊓ bᶜ :=
+  compl_sup_distrib _ _
+#align compl_sup compl_sup
+
+theorem compl_le_himp : aᶜ ≤ a ⇨ b :=
+  (himp_bot _).ge.trans <| himp_le_himp_left bot_le
+#align compl_le_himp compl_le_himp
+
+theorem compl_sup_le_himp : aᶜ ⊔ b ≤ a ⇨ b :=
+  sup_le compl_le_himp le_himp
+#align compl_sup_le_himp compl_sup_le_himp
+
+theorem sup_compl_le_himp : b ⊔ aᶜ ≤ a ⇨ b :=
+  sup_le le_himp compl_le_himp
+#align sup_compl_le_himp sup_compl_le_himp
+
+-- `p → ¬ p ↔ ¬ p`
+@[simp]
+theorem himp_compl (a : α) : a ⇨ aᶜ = aᶜ := by rw [← himp_bot, himp_himp, inf_idem]
+#align himp_compl himp_compl
+
+-- `p → ¬ q ↔ q → ¬ p`
+theorem himp_compl_comm (a b : α) : a ⇨ bᶜ = b ⇨ aᶜ := by simp_rw [← himp_bot, himp_left_comm]
+#align himp_compl_comm himp_compl_comm
+
+theorem le_compl_iff_disjoint_right : a ≤ bᶜ ↔ Disjoint a b := by
+  rw [← himp_bot, le_himp_iff, disjoint_iff_inf_le]
+#align le_compl_iff_disjoint_right le_compl_iff_disjoint_right
+
+theorem le_compl_iff_disjoint_left : a ≤ bᶜ ↔ Disjoint b a :=
+  le_compl_iff_disjoint_right.trans Disjoint.comm
+#align le_compl_iff_disjoint_left le_compl_iff_disjoint_left
+
+theorem le_compl_comm : a ≤ bᶜ ↔ b ≤ aᶜ := by
+  rw [le_compl_iff_disjoint_right, le_compl_iff_disjoint_left]
+#align le_compl_comm le_compl_comm
+
+alias le_compl_iff_disjoint_right ↔ _ Disjoint.le_compl_right
+
+alias le_compl_iff_disjoint_left ↔ _ Disjoint.le_compl_left
+
+alias le_compl_comm ← le_compl_iff_le_compl
+
+alias le_compl_comm ↔ le_compl_of_le_compl _
+
+theorem disjoint_compl_left : Disjoint (aᶜ) a :=
+  disjoint_iff_inf_le.mpr <| le_himp_iff.1 (himp_bot _).ge
+#align disjoint_compl_left disjoint_compl_left
+
+theorem disjoint_compl_right : Disjoint a (aᶜ) :=
+  disjoint_compl_left.symm
+#align disjoint_compl_right disjoint_compl_right
+
+theorem LE.le.disjoint_compl_left (h : b ≤ a) : Disjoint (aᶜ) b :=
+  disjoint_compl_left.mono_right h
+#align has_le.le.disjoint_compl_left LE.le.disjoint_compl_left
+
+theorem LE.le.disjoint_compl_right (h : a ≤ b) : Disjoint a (bᶜ) :=
+  disjoint_compl_right.mono_left h
+#align has_le.le.disjoint_compl_right LE.le.disjoint_compl_right
+
+theorem IsCompl.compl_eq (h : IsCompl a b) : aᶜ = b :=
+  sorry
+  -- Hopefully will work in nightly-2022-11-30
+  -- h.1.le_compl_left.antisymm' <| Disjoint.le_of_codisjoint disjoint_compl_left h.2
```

Comment:
Doesn't seem like it

Reply:
This is a very mysterious 
```
failed to synthesize instance
  BoundedOrder α
```
I can't make it into an MWE

Reply:
Here is an MWE, I think:
```lean
variable {α : Type _}

class BoundedOrder (α : Type u) [LE α]

class DistribLattice (α) extends LE α 

def Disjoint [DistribLattice α] (a b : α) : Prop := sorry

def Codisjoint [DistribLattice α] (a b : α) : Prop := sorry

theorem Disjoint.le_of_codisjoint [DistribLattice α] [BoundedOrder α] {a b c : α}
    (hab : Disjoint a b) (hbc : Codisjoint b c) : a ≤ c :=
  sorry

class HeytingAlgebra (α : Type _) extends LE α

instance (priority := 100) HeytingAlgebra.toDistribLattice [HeytingAlgebra α] : DistribLattice α :=
  sorry

instance (priority := 100) HeytingAlgebra.toBoundedOrder [HeytingAlgebra α] : BoundedOrder α :=
  sorry

example [HeytingAlgebra α] {a b c : α} (h₁ : Disjoint a b) (h₂ : Codisjoint b c) : a ≤ c :=
  Disjoint.le_of_codisjoint h₁ h₂
```

Reply:
Changing the `sorry` for both of your instances into `{  }` seems to fix the problem in your MWE.

Reply:
Then probably I simplified too much, sorry!